### DISCUSSION TITLE
Add Git Bash (MINGW)

### DISCUSSION THREAD

**Discussion by Gebhartj** [Taitava/obsidian-shellcommands#208] [Ideas - done & released]
I would like to use the same Shell across operating systems. I have Git installed on all computers anyway.

**Comment by Taitava** [comment]
Yea, I agree. I need to think about when I can implement support for custom shells (#108). When that support gets implemented, you can then add Git Bash to the list of selectable shells.

There are certain properties that users need to define for each custom shell:
 - How special characters are escaped. There are currently two escapers defined in the _Shell commands_ plugin: `ShEscaper` (suitable for Git Bash) and `PowerShellEscaper`. Users must pick one for their custom shell.
 - What separator character to use between `PATH` entries (#150 and #204). I guess this is more dependent on the operating system, not actually on the shell, so I might leave this definition off from the custom shell settings.

To sum it up, my point is that I don't know yet if there will be new features in the future that will need some new details from the currently used shell. That's why I haven't implemented the custom shell support yet. But maybe I'll just implement it, and if there later comes new features that require additional details, then users just need to supplement the custom shell definitions after version upgrades.

**Reply by Taitava** [reply]
@Gebhartj if you are still interested in using Git Bash with this plugin, you can participate in [`0.19.0-beta.1` test](https://github.com/Taitava/obsidian-shellcommands/discussions/108#discussioncomment-5277523).

Once you have installed the beta version, you can follow [these instructions on how to define custom shell settings](https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Settings+for+custom+shells). For MinGW-w64, you'll need to use the following settings:

- _Executable binary file path_: `C:\Program Files\Git\bin\bash.exe` (assuming this is where it's installed on your machine)
- _Host operating system_: Windows
- _Windows: Quote shell arguments_: On
- _Shell's operating system_: Linux
- _Special characters escaping_: `Unix shell style with \ as escape character`
- _Path translator_:
    E.g. this should work well:
    ```
    const pathParts = absolutePath.match(/^([a-z]):\/(.+)$/ui);
    const driveLetter = pathParts[1];
    const trailingPath = pathParts[2];
    return '/' + driveLetter.toLocaleLowerCase() + '/' + trailingPath;
    ```
- _Wrapper for shell command_: Can be left empty.
- _Execute a command to test the shell_:
    Can be for example:
    ```bash
    echo "Hello, current working directory is: $(pwd)"
    ```

Please let me know if the custom shell feature works for you, or if you have any questions 🙂👍.

**Reply by Taitava** [reply]
I just wrote a documentation page with the same MinGW-w64 settings: https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Windows+-+MinGW-w64

The settings are the same as above, just a bit more additional information.

**Comment by Taitava** [comment]
# Moving this dicussion from _Ideas_ to _Ideas - done & released_
The _Custom shells_ feature I mentioned above has been released some time ago in `0.19.0`, so this feature request should be fulfilled now, as far as I know.

If anyone still has any MINGW related questions, they can still be asked in this discussion.
